[WIP] Remove codegen, implement registers for sc-keynodes and sc-agents (#124)
ostis-ai/sc-machine: docs/sc-memory/api/cpp/extended/agents/agents.md

Reviewer: there is no info about m_context?

Author: Fixed

# **C++ Agents API**

!!! warning
    This documentation is correct for only versions of sc-machine that >= 0.10.0.
--- 

This API provides functionality to implement agents on C++.

## **What is the agent-driven model?**

> The sc-machine implements the **agent-driven model** to process information. In the agent-based model, agents exchange messages only through shared memory, adding a new agent or eliminating one or more existing agents does not result in changes to other agents, agent initiation is decentralized and most often independent of each other. 

Author: I don't agree with you. Let's discuss

## **What does agent specification represent?**

All knowledge about an agent: *primary initiation condition*, *class of actions* it can interpret, *initiation condition*, and *result condition*, are part of **agent's specification**. This specification can be represented either in a knowledge base, using SC-code, or programly, using sc-machine API.

Let's describe specification for abstract sc-agent of counting power of specified set in SCs-code and SCg-code. An abstract sc-agent is a class of functionally equivalent agents, different instances of which can be implemented in different ways. Each abstract sc-agent has a specification corresponding to it.

```scs
// Abstract sc-agent
agent_for_calculating_set_power
<- abstract_sc_agent;
=> nrel_primary_initiation_condition: 
    // Class of sc-event and listen (subscription) sc-element
    (sc_event_add_outgoing_arc => action_initiated); 
=> nrel_sc_agent_action_class:
    // Class of actions to be interpreted by agent

>     action_for_calculating_set_power; 

Reviewer 2: ```suggestion
    action_calculate_set_power; 
```

=> nrel_initiation_condition_and_result: 
    (..agent_for_calculating_set_power_initiation_condition 
        => ..agent_for_calculating_set_power_result_condition);
<= nrel_sc_agent_key_sc_elements:
// Set of key sc-elements used by this agent
{
    action_initiated;
    action;
    action_for_calculating_set_power;
    nrel_set_power;
};
=> nrel_inclusion: 
    // Instance of abstract sc-agent; concrete implementation of agent in C++
    agent_for_calculating_set_power_implementation 
    (*
        <- platform_dependent_abstract_sc_agent;;
        // Set of links with paths to sources of agent programs
        <= nrel_sc_agent_program: 
        {
            [github.com/path/to/agent/sources] 

>             (* => nrel_format: format_github_source_link;; *)

Author: Now it is not handled by sc-machine and sc-component-manager. I'll describe it in this line

        };; 
    *);;

// Full initiation condition of agent
..agent_for_calculating_set_power_initiation_condition
= [*
    action_for_calculating_set_power _-> .._action;;
    action_initiated _-> .._action;;
    .._action _-> rrel_1:: .._parameter;;
*];; 
// Agent should check by this template that initiated action is instance of 
// class `action_for_calculating_set_power` and that it has argument.

// Full result condition of agent
..agent_for_calculating_set_power_result_condition
= [*
    action_for_calculating_set_power _-> .._action;;
    action_initiated _-> .._action;;
    action_finished_successfully _-> .._action;;
    .._action _-> rrel_1:: .._parameter;;
    .._action _=> nrel_result:: .._result;;
*];;
// Agent should check by this template that initiated action is finished 
// and that it has result.
```

> <image src="../images/agents/agent_for_calculating_set_power_specification.png"></image>

Reviewer 2: nrel_format is crooked. change direction of nrel_sc_agent_program

## **What are ways of providing the agent's specification?**

The sc-machine API provides two types of functionalities to implement an agent in C++:

> a. when the agent's specification is represented in the knowledge base and used by the agent;

Reviewer 2: space

b. when the agent's specification is represented in C++ code and also used by the agent.

In both cases, the agent's specification can be static, dynamic, or semi-dynamic.

1. **Static agent specification** is a specification provided externally in the agent's class (via overriding public getters). This specification isn't saved in the knowledge base because changes in the knowledge base can't automatically alter user-specified code. To provide this type of agent specification, you can use overriding public getters in your agent's class (see documentation below).

> 2. **Dynamic agent specification** is a specification provided in the knowledge base or initially in the code but automatically saved into the knowledge base. After that, the first agent changes the specification of the second agent, and the second agent uses the changed specification. To implement this type of specification, use the API of `ScModule` class and the API of `ScAgentBuilder` class (see C++ Modules).

Reviewer 2: ```suggestion
2. **Dynamic agent specification** is a specification provided in the knowledge base or initially in the code but automatically saved into the knowledge base. After that, the first agent changes the specification of the second agent, and the second agent uses the changed specification. To implement this type of specification, use the API of `ScModule` class and the API of `ScAgentBuilder` class (see [**C++ Modules API**](modules.md)).
```

3. **Semi-dynamic agent specification** can be obtained when the specification is provided in the knowledge base or initially in the code and appended externally (via overriding public getters).

Static agent specification can be useful if you are implementing an agent in C++ for the first time or if you want to minimize the number of searches in the knowledge base. Dynamic agent specification provides the opportunity to analyze and change this specification by other agents. Semi-dynamic specification can be useful if you want to change some parts of this specification but still want the agent's calls to be quick.

---

## **Static agent specification**

Here, the API for implementing an agent with a static specification will be discussed. This is the easiest implementation variant to understand for you. To learn about dynamic agent specification, see [**C++ Modules API**](modules.md).

There are two main classes that you can use to implement an agent: `ScAgent` and `ScActionAgent`.

### **ScAgent**

This class can be used for all classes of agents. The example using this class is represented below.

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

> // Override your agent class from `ScAgent` class and specify template argument 

Reviewer 2: ```suggestion
// Inherit your agent class from `ScAgent` class and specify template argument 
```

// as sc-event class. Here `ScEventAddIncomingArc<ScType::EdgeAccessConstPosPerm>` 
// is type of event to which the given agent reacts.

> class MyAgent : public ScAgent<ScEventAddIncomingArc<ScType::EdgeAccessConstPosPerm>>

Author: Now, agents can trigger on any type of events

> class MyAgent : public ScAgent<ScEventAddIncomingArc<ScType::EdgeAccessConstPosPerm>>

Author: Yes, you can override one of methods that check initiation condition and describe this check or implement initiation condition template in knowledge base

{
public:
  // Here you should specify class of actions which the given agent interpreters. 
  // This overriding is required.
  ScAddr GetActionClass() const override;
  // Here you should implement program of the given agent. 
  // This overriding is required.
  ScResult DoProgram(
    ScEventAddIncomingArc<ScType::EdgeAccessConstPosPerm> const & event, 

>     ScAction & action) override;

Reviewer 2: Add DoProgram without `event` argument

  // Other user-defined methods.
};
```

See [**C++ Events API**](events.md) and [**C++ Event subscriptions API**](event_subscriptions.md) to learn how to use and handle sc-events.

!!! note
    Define action class as keynode in agent class or keynodes class.

!!! warning
    You must override methods `GetActionClass` and `DoProgram`.

You can specify any existing event types as a template argument to the `ScAgent` class. For example, you can create agent that will be triggered to sc-event of removing sc-element.

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

class MyAgent : public ScAgent<ScEventEraseElement>
{
public:
  ScAddr GetActionClass() const override;
  ScResult DoProgram(
    ScEventEraseElement const & event, ScAction & action) override;

  // Other user-defined methods.
};
```

> If you want to change specification of this agent in kb, then write like this:

Reviewer 2: ```suggestion
If you want to change specification of this agent in knowledge base, then write like this:
```

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

class MyAgent : public ScBaseAgent
{
public:
  ScAddr GetActionClass() const override;
  ScResult DoProgram(
    ScElementaryEvent const & event, ScAction & action) override;

  // Other user-defined methods.
};
```

This implementation allows to provide any sc-event type to `DoProgram`.

!!! note

>     `MyAgent` is alias for ScAgent<ScElementaryEvent>`.

Reviewer 2: ```suggestion
    `ScBaseAgent` is alias for `ScAgent<ScElementaryEvent>`.
```

!!! note
    If you provide specification of your agent in knowledge base, then you don't need to override `GetActionClass`.

### **ScActionAgent**

In addition to agents that initiate actions themselves and then interpret these actions, there is a need to implement agents that interpret actions initiated by other agents. For this class of agents, it is much easier to create a initial initiation construction in the knowledge base.

This class can be only used for agents that should be triggered to sc-event of adding output sc-arc from `action_initiated` class. These agents are named action agents. Action agent interpreters action initiated by other agent.

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

> // Override your agent class from `ScAgent` class and specify template argument 

Author: ScAgent has template argument -- ScEventGenerateIncomingArc<ScType::EdgeAccessConstPosPerm>

// as sc-event class. Here `ScActionEvent` is type of event to which 
// the given agent reacts.
class MyAgent : public ScActionAgent
{
public:
  // Here you should specify class of actions which the given agent interpreters. 
  // This overriding is required.
  ScAddr GetActionClass() const override;
  // Here you should implement program of the given agent. 
  // This overriding is required.
  ScResult DoProgram(ScActionEvent const & event, ScAction & action) override;

  // Other user-defined methods.
};
```

!!! note
    `ScActionAgent` has default `GetInitiationConditionTemplate` that returns template that can be used to check that initiated action is action with class of specified agent.

!!! note
    `ScActionEvent` is alias for `ScEventAddOutgoingArc<ScType::EdgeAccessConstPosPerm>` with subscription sc-element `action_initiated`.

---

### **ScAgentAbstract**

There is a base class for agents in C++. This class provides implemented methods to retrieve elements of the agent's specification from the knowledge base. All these methods can be overridden in your agent class.

> This class does not provide methods to subscribe and unsubscribe a specified agent. This functionality is provided by the classes `ScAgent` and `ScActionAgent`. You should derive your agent class from these classes, not from the ScAgentAbstract class (see documentation above).

Author: subscribe and unsubscribe agent

---

### **Required methods**

#### **GetActionClass**

> It gets action class that the agent of specified class interpreters. If the abstract sc-agent for this agent class does not have an action class, then method will throw `utils::ExceptionInvalidState`.

Author: It is a set of implementations of agent

{
  // You must specify valid sc-address of action class. 
  // In other case, the given sc-agent can’t be registered in sc-memory.
  return MyKeynodes::my_action;
}
```

See [**C++ Keynodes API**](keynodes.md) and learn how to define keynodes and use them for your agent.

#### **DoProgram**

In the agent's program lies its basic operating logic. Using program, the agent processes an input construction and generates an output construction.
Each agent interprets action with the help of its program.

```cpp
ScResult MyAgent::DoProgram(ScActionEvent const & event, ScAction & action)
{
  // Class `ScAction` encapsulates information about sc-action. The provided 
  // action is action that the given agent interpreters right now. 
  // It belongs to `MyKeynodes::my_action` class. 
  // Actions are copyable and movable. `ScAction` is derived from `ScAddr`.
 
  // Implement logic of your agent...

  // You must specify action state in all ends of your agent program. 
  // `FinishSuccessfully` sets action as `action_finished_successfully`.
  // You can’t create object of `ScResult` via constructor, because it is 
  // private.
  return action.FinishSuccessfully(); 
}
```

!!! warning

>     If you don't catch sc-exceptions in `DoProgram` then sc-machine will catch them and will warn you about it.

Author: Let's discuss with all.

I propose to add action to `action_finished_with_error` and to `action_finished_forcibly`

>     If you don't catch sc-exceptions in `DoProgram` then sc-machine will catch them and will warn you about it.

Author: Fixed

  return ScKeynodes::action_initiated;
}
```

!!! warning
    Don't override `GetEventClass` and `GetEventSubscriptionElement` for agent with statically specified sc-event type. Your code can't be compiled. Override them, if your agent class derive `ScAgent<ScElementaryEvent>` (`ScBaseAgent`).

#### **GetInitiationCondition**

It gets initiation condition for agent of this class. This method will throw `utils::ExceptionInvalidState` if the abstract sc-agent for this agent class does not have initiation condition.

You can redefine this method in your agent class.

```cpp
ScAddr MyAgent::GetInitiationCondition() const
{
  // You must specify valid sc-address of initiation condition. 
  // In other case, the given sc-agent can’t be called.
  return ScKeynodes::my_initiation_condition;
}
```

#### **GetInitiationConditionTemplate**

You can specify the initiation condition template in code rather than in the knowledge base.

```cpp
ScTemplate MyAgent::GetInitiationConditionTemplate() const
{
  ScTemplate initiationCondition;
  // You must specify in template of initiation condition 
  // a triple in place of which you can substitute initiated 

>   // event, otherwise your agent will be called even when 

Author: See `sc_agent.tpp`

>   // event, otherwise your agent will be called even when 

Author: Fixed

// You can use all opportunities of sc-memory API to check initiation 
// condition of your agent.
```

!!! note
    Each of these methods will be called before `DoProgram` call.

!!! warning
    You can specify only one of methods: `GetInitiationCondition`, `GetInitiationConditionTemplate` and `CheckInitiationCondition`. Otherwise, your code can't be compiled.

#### **GetResultCondition**

It gets result condition for agent of this class. This method will throw `utils::ExceptionInvalidState` if the abstract sc-agent for this agent class does not have result condition.

You can redefine this method in your agent class.

```cpp
ScAddr MyAgent::GetResultCondition() const
{
  // You must specify valid sc-address of result condition. 
  // In other case, the given sc-agent can’t be finished successfully.
  return ScKeynodes::my_result_condition;
}
```

#### **GetResultConditionTemplate**

You can specify the result condition template in code rather than in the knowledge base.

```cpp
ScTemplate MyAgent::GetResultConditionTemplate() const
{
  ScTemplate resultCondition;
  // You must specify in template of result condition 
  // a triple in place of which you can substitute initiated 

>   // event, otherwise your agent will be finished even when 

Author: What's wrong?

>   // event, otherwise your agent will be finished even when 

Author: Fixed

  // result condition is fulfilled for the action of 
  // another agent.
  resultCondition.Triple(
    ScKeynodes::action_initiated,
    ScType::EdgeAccessVarPosPerm,
    ScType::NodeVar >> "_action"
  );
  // After specify triples that only apply to your agent.
  resultCondition.Triple(
    MyKeynodes::my_action,
    ScType::EdgeAccessVarPosPerm,
    "_action"
  );
  resultCondition.Quintuple(
    "_action",
    ScType::EdgeDCommonVar,
    ScType::NodeVarStruct,
    ScType::EdgeAccessVarPosPerm,
    ScKeynodes::nrel_result
  );
  return resultCondition;
}
```

#### **CheckResultCondition**

For speed, you can implement the agent result condition in the form of checks on iterators.

```cpp
sc_bool MyAgent::CheckResult(ScActionEvent const & event, ScAction & action)
{
  return m_memoryCtx.HelperCheckEdge(
    ScType::EdgeAccessConstPosPerm, action.GetResult(), MyKeynodes::my_class);

> }

Author: Fixed

You can specify any existing event types as a template argument to the `ScAgent` class. For example, you can create agent that will be triggered to sc-event of removing sc-element.

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

class MyAgent : public ScAgent<ScEventEraseElement>
{
public:
  ScAddr GetActionClass() const override;
  ScResult DoProgram(
    ScEventEraseElement const & event, ScAction & action) override;

  // Other user-defined methods.
};
```

If you want to change specification of this agent in knowledge base, then write like this:

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

class MyAgent : public ScBaseAgent
{
public:
  ScAddr GetActionClass() const override;
  ScResult DoProgram(
    ScElementaryEvent const & event, ScAction & action) override;

  // Other user-defined methods.
};
```

This implementation allows to provide any sc-event type to `DoProgram`.

!!! note
    `ScBaseAgent` is alias for `ScAgent<ScElementaryEvent>`.

!!! note
    If you provide specification of your agent in knowledge base, then you don't need to override `GetActionClass`.

### **ScActionAgent**

> In addition to agents that initiate actions themselves and then interpret these actions, there is a need to implement agents that interpret actions initiated by other agents. For this class of agents, it is much easier to create a initial initiation construction in the knowledge base.
> 
> This class can be only used for agents that should be triggered to sc-event of generating output sc-arc from `action_initiated` class. These agents are named action agents. Action agent interpreters action initiated by other agent.

Reviewer 3: ```suggestion
In multi-agent systems most of the agents are implemented to execute actions initiated by other agents. While `ScAgent` is useful to create broad event handling logic, using it to handle action initiations requires some boilerplate. We've implemented another agent class to make it easier for our users to implement action-executing agents. These agents are named action agents. Implementing these agents requires passing action class node rather than checking initiation condition manually.

This class can be only used for agents that should be triggered by generating an output sc-arc from `action_initiated` class node.
```

> This class can be only used for agents that should be triggered to sc-event of generating output sc-arc from `action_initiated` class. These agents are named action agents. Action agent interpreters action initiated by other agent.

Author: Not actual

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

// Inherit your agent class from `ScAgent` class and specify template argument 
// as sc-event class. Here `ScActionEvent` is type of event to which 
// the given agent reacts.
class MyAgent : public ScActionAgent
{
public:

>   // Here you should specify class of actions which the given agent interpreters. 

Author: Not actual

---

### **Required methods**

#### **GetActionClass**

It gets action class that the agent of specified class interpreters. If the abstract sc-agent for this agent class does not have an action class, then method will throw `utils::ExceptionInvalidState`.

```cpp
// File my_agent.cpp
#include "my_agent.hpp"

#include "keynodes/my_keynodes.hpp"

ScAddr MyAgent::GetActionClass() const
{
  // You must specify valid sc-address of action class. 
  // In other case, the given sc-agent can’t be registered in sc-memory.
  return MyKeynodes::my_action;
}
```

See [**C++ Keynodes API**](keynodes.md) and learn how to define keynodes and use them for your agent.

#### **DoProgram**

In the agent's program lies its basic operating logic. Using program, the agent processes an input construction and generates an output construction.
Each agent interprets action with the help of its program.

```cpp
ScResult MyAgent::DoProgram(ScActionEvent const & event, ScAction & action)
{
  // Class `ScAction` encapsulates information about sc-action. The provided 

>   // action is action that the given agent interpreters right now. 

Author: Not actual

{
  // Some logic...
 
  action.FormResult(foundAddr1, generatedAddr1, ...); 
  // Or you can use UpdateResult.
  return action.FinishSuccessfully();
}
```

```cpp
ScResult MyAgent::DoProgram(ScActionEvent const & event, ScAction & action)
{
  // Some logic...
 
  action.SetResult(structureAddr);
  return action.FinishSuccessfully();
}
```

##### Handling action finish state

```cpp
ScResult MyAgent::DoProgram(ScActionEvent const & event, ScAction & action)
{
  // Some logic...

  if (/* case 1 */)
    return action.FinishSuccessfully();
  else if (/* case 2 */)
    return action.FinishUnsuccessfully();
  else
    return action.FinishWithError();
}
```

```cpp
ScResult MyAgent::Program(MyEventType const & event, ScAction & action)
{
  action.IsInitiated(); // result: SC_TRUE
  action.IsFinished(); // result: SC_FALSE
  action.IsFinishedSuccessfully(); // result: SC_FALSE

  // Some logic...

  return action.FinishSuccessfully();
}
```

!!! warning
    The API of `ScAction` provides other methods. Not use `GetResult` for initiated but not finished action and `Initiate` for initiated or finished action.

---

### **Not required methods**

#### **GetAbstractAgent**

> It searches abstract agent for agent of specified class. If the agent implementation for this agent class is not included in any abstract sc-agent, then `GetAbstractAgent` will throw `utils::ExceptionInvalidState`.

Author: I'll do this later

> It searches abstract agent for agent of specified class. If the agent implementation for this agent class is not included in any abstract sc-agent, then `GetAbstractAgent` will throw `utils::ExceptionInvalidState`.

Author: Fixed

# **C++ Agents API**

!!! warning
    This documentation is correct for only versions of sc-machine that >= 0.10.0.
--- 

This API provides functionality to implement agents on C++.

## **What is the agent-driven model?**

> The sc-machine implements the **agent-driven model** to process information. In the agent-based model, agents exchange messages only through shared memory, adding a new agent or eliminating one or more existing agents does not result in changes to other agents, agent initiation is decentralized and most often independent of each other. 

Author: Agree with you

=> nrel_initiation_condition_and_result: 
    (..agent_calculate_set_power_initiation_condition 
        => ..agent_calculate_set_power_result_condition);
<= nrel_sc_agent_key_sc_elements:
// Set of key sc-elements used by this agent
{
    action_initiated;
    action;
    action_calculate_set_power;
    nrel_set_power;
};
=> nrel_inclusion: 
    // Instance of abstract sc-agent; concrete implementation of agent in C++
    agent_calculate_set_power_implementation 
    (*
        <- platform_dependent_abstract_sc_agent;;
        // Set of links with paths to sources of agent programs
        <= nrel_sc_agent_program: 
        {
            [github.com/path/to/agent/sources] 
            (* => nrel_format: format_github_source_link;; *)
        };; 
    *);;

// Full initiation condition of agent
..agent_calculate_set_power_initiation_condition
= [*
    action_calculate_set_power _-> .._action;;
    action_initiated _-> .._action;;
    .._action _-> rrel_1:: .._parameter;;
*];; 
// Agent should check by this template that initiated action is instance of 
// class `action_calculate_set_power` and that it has argument.

// Full result condition of agent

> ..agent_calculate_set_power_result_condition

Author: Without specifyinп triplet in the result condition, we cannot substitute the correct action into this pattern. We need specify that `_action` should be belong to `action_initiated`

> ..agent_calculate_set_power_result_condition

Author: I agree with the second part of your proposition

> ..agent_calculate_set_power_result_condition

Author: Fixed

= [*
    action_calculate_set_power _-> .._action;;
    action_initiated _-> .._action;;
    action_finished_successfully _-> .._action;;
    .._action _-> rrel_1:: .._parameter;;
    .._action _=> nrel_result:: .._result;;
*];;
// Agent should check by this template that initiated action is finished 
// and that it has result.
```

<image src="../images/agents/agent_calculate_set_power_specification.png"></image>

## **What are ways of providing the agent's specification?**

The sc-machine API provides two types of functionalities to implement an agent in C++:

a. when the agent's specification is represented in the knowledge base and used by the agent;

> b. when the agent's specification is represented in C++ code and also used by the agent.

Reviewer 4: the list is not nicely formatted and is hard to read

3. **Semi-dynamic agent specification** can be obtained when the specification is provided in the knowledge base or initially in the code and appended externally (via overriding public getters).

Static agent specification can be useful if you are implementing an agent in C++ for the first time or if you want to minimize the number of searches in the knowledge base. Dynamic agent specification provides the opportunity to analyze and change this specification by other agents. Semi-dynamic specification can be useful if you want to change some parts of this specification but still want the agent's calls to be quick.

---

## **Static agent specification**

Here, the API for implementing an agent with a static specification will be discussed. This is the easiest implementation variant to understand for you. To learn about dynamic agent specification, see [**C++ Modules API**](modules.md).

There are two main classes that you can use to implement an agent: `ScAgent` and `ScActionAgent`.

### **ScAgent**

> This class can be used for all classes of agents. The example using this class is represented below.

Reviewer 4: what do you mean by `all classes of agents`? Only 2 agent classes were mentioned on this page: platform-independent and platform-dependent. It can confuse users

!!! note
    Define action class as keynode in agent class or keynodes class.

!!! warning
    You must override methods `GetActionClass` and `DoProgram`.

You can specify any existing event types as a template argument to the `ScAgent` class. For example, you can create agent that will be triggered to sc-event of removing sc-element.

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

class MyAgent : public ScAgent<ScEventEraseElement>
{
public:
  ScAddr GetActionClass() const override;
  ScResult DoProgram(
    ScEventEraseElement const & event, ScAction & action) override;

  // Other user-defined methods.
};
```

If you want to change specification of this agent in knowledge base, then write like this:

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

class MyAgent : public ScBaseAgent
{
public:
  ScAddr GetActionClass() const override;
  ScResult DoProgram(
    ScElementaryEvent const & event, ScAction & action) override;

  // Other user-defined methods.
};
```

This implementation allows to provide any sc-event type to `DoProgram`.

!!! note
    `ScBaseAgent` is alias for `ScAgent<ScElementaryEvent>`.

!!! note

>     If you provide specification of your agent in knowledge base, then you don't need to override `GetActionClass`.

Author: I can write an algorithm for loading a specification for a given agent. I am afraid that it should not be difficult for the user

>     If you provide specification of your agent in knowledge base, then you don't need to override `GetActionClass`.

Reviewer 4: I'm talking about something a little different. It's necessary to clarify how to put the specification in the DB so that in the C++ implementation of the agent it would be clear that this is the specification of this agent.

>     If you provide specification of your agent in knowledge base, then you don't need to override `GetActionClass`.

Author: I will leave link to ScModule API here and write guide for writing agent with specification in knowledge base in separate PR

>     If you provide specification of your agent in knowledge base, then you don't need to override `GetActionClass`.

Author: Fixed

### **ScActionAgent**

> In multi-agent systems most of the agents are implemented to execute actions initiated by other agents. While `ScAgent` is useful to create broad event handling logic, using it to handle action initiations requires some boilerplate. We've implemented another agent class to make it easier for our users to implement action-executing agents. These agents are named action agents. Implementing these agents requires passing action class node rather than checking initiation condition manually.

Author: We can add a link to our guide there

> In multi-agent systems most of the agents are implemented to execute actions initiated by other agents. While `ScAgent` is useful to create broad event handling logic, using it to handle action initiations requires some boilerplate. We've implemented another agent class to make it easier for our users to implement action-executing agents. These agents are named action agents. Implementing these agents requires passing action class node rather than checking initiation condition manually.

Author: Fixed

This class can be only used for agents that should be triggered by generating an output sc-arc from `action_initiated` class node.

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

// Inherit your agent class from `ScActionAgent` class.
class MyAgent : public ScActionAgent
{
public:
  // Here you should specify class of actions which the given agent performs. 
  // This overriding is required.

>   ScAddr GetActionClass() const override;

Author: GetActionClass here overrides GetActionClass in ScAgent

>   ScAddr GetActionClass() const override;

Reviewer 4: Yes, I see but my point about different semantic of action class here and action class in ScAgent

  // Here you should implement program of the given agent. 
  // This overriding is required.
  ScResult DoProgram(ScActionEvent const & event, ScAction & action) override;
  // Here `ScActionEvent` is type of event to which 
  // the given agent reacts.

  // Other user-defined methods.
};
```

!!! note

>     `ScActionAgent` has default `GetInitiationConditionTemplate` that returns template that can be used to check that initiated action is action with class of specified agent.

Author: We discussed it. We decided that we should leave it as it is

}
```

See [**C++ Keynodes API**](keynodes.md) and learn how to define keynodes and use them for your agent.

#### **DoProgram**

This method is executed when agent checked initiation condition successfully. Using program, the agent processes an input construction and generates an output construction.
Each agent performs action with the help of its program.

```cpp
ScResult MyAgent::DoProgram(ScActionEvent const & event, ScAction & action)
{
  // Class `ScAction` encapsulates information about sc-action. The provided 
  // action is action that the given agent performs right now. 
  // It belongs to `MyKeynodes::my_action` class. 
  // Actions are copyable and movable. `ScAction` is inherited from `ScAddr`.

  // `ScActionEvent` class is event type on which the given agent 
  // is triggered. It is encapsulate information about sc-event. 
  // The provided event is event on which the agent is triggered right now. 
  // It has methods to get information about initiated sc-event: `GetUser`, 
  // `GetGeneratedArc`, `GetSubscriptionElement`, `GetArcSourceElement`.
 
  // Implement logic of your agent...

  // You must specify action state in all ends of your agent program. 

>   // `FinishSuccessfully` sets action as `action_finished_successfully`.

Author: I'll do it a little later

>   // `FinishSuccessfully` sets action as `action_finished_successfully`.

Author: Fixed

{
  // Some logic...
 
  action.FormResult(foundAddr1, generatedAddr1, ...); 
  // Or you can use UpdateResult.
  return action.FinishSuccessfully();
}
```

```cpp
ScResult MyAgent::DoProgram(ScActionEvent const & event, ScAction & action)
{
  // Some logic...
 
  action.SetResult(structureAddr);
  return action.FinishSuccessfully();
}
```

##### Handling action finish state

```cpp
ScResult MyAgent::DoProgram(ScActionEvent const & event, ScAction & action)
{
  // Some logic...

  if (/* case 1 */)
    return action.FinishSuccessfully();
  else if (/* case 2 */)
    return action.FinishUnsuccessfully();
  else
    return action.FinishWithError();
}
```

```cpp
ScResult MyAgent::Program(MyEventType const & event, ScAction & action)
{
  action.IsInitiated(); // result: SC_TRUE
  action.IsFinished(); // result: SC_FALSE
  action.IsFinishedSuccessfully(); // result: SC_FALSE

  // Some logic...

  return action.FinishSuccessfully();
}
```

!!! warning
    The API of `ScAction` provides other methods. Don't use `GetResult` for initiated but not finished action and `Initiate` for initiated or finished action.

---

### **Not required methods**

> #### **GetAbstractAgent**

Author: Renamed ScAgentAbstract to ScAgentBase

You can redefine this method in your agent class.

```cpp
ScAddr MyAgent::GetEventSubscriptionElement() const
{
  // You must specify valid sc-address of sc-event subscription sc-element. 
  // In other case, the given sc-agent can’t be registered in sc-memory.
  return ScKeynodes::action_initiated;
}
```

!!! warning
    Don't override `GetEventClass` and `GetEventSubscriptionElement` for agent with statically specified sc-event type. Your code can't be compiled. Override them, if your agent class inherit `ScAgent<ScElementaryEvent>` (`ScBaseAgent`).

#### **GetInitiationCondition**

It gets initiation condition for agent of this class. This method will throw `utils::ExceptionInvalidState` if the abstract sc-agent for this agent class does not have initiation condition.

You can redefine this method in your agent class.

```cpp
ScAddr MyAgent::GetInitiationCondition() const
{
  // You must specify valid sc-address of initiation condition. 
  // In other case, the given sc-agent can’t be called.
  return ScKeynodes::my_initiation_condition;
}
```

> #### **GetInitiationConditionTemplate**

Author: You can use `GetInitiationCondition` and `GetInitiationConditionTemplate` as getters.

> #### **GetInitiationConditionTemplate**

Reviewer 4: Need to discuss

    ScKeynodes::action_initiated,
    ScType::EdgeAccessVarPosPerm,
    ScType::NodeVar >> "_action"
  );
  // After specify triples that only apply to your agent.
  initiationCondition.Triple(
    MyKeynodes::my_action,
    ScType::EdgeAccessVarPosPerm,
    "_action"
  );
  return initiationCondition;
}
```

#### **CheckInitiationCondition**

For speed, you can implement the agent initiation condition in the form of checks on iterators.

```cpp
sc_bool MyAgent::CheckInitiationCondition(ScActionEvent const & event)
{
 // ScActionEvent is event type on which the given agent triggered. 
 // It is encapsulate information about sc-event. The provided event 
 // is event on which the agent is triggered  right now. It has methods 
 // to get information about initiated sc-event: GetUser, GetGeneratedArc, 
 // GetSubscriptionElement, GetArcSourceElement, GetArcTargetElement.
 // All events are not copyable and movable.
 return m_memoryCtx.HelperCheckEdge(
   ScType::EdgeAccessConstPosPerm, 
   MyKeynodes::my_action, 
   event.GetArcTargetElement());
}
// You can use all opportunities of sc-memory API to check initiation 
// condition of your agent.
```

!!! note

>     Each of these methods will be called before `DoProgram` call.

Author: I don't want to make a large nesting of sections

{
    action_initiated;
    action;
    action_calculate_set_power;
    nrel_set_power;
};
=> nrel_inclusion: 
    // Instance of abstract sc-agent; concrete implementation of agent in C++
    agent_calculate_set_power_implementation 
    (*
        <- platform_dependent_abstract_sc_agent;;
        // Set of links with paths to sources of agent programs
        <= nrel_sc_agent_program: 
        {
            [github.com/path/to/agent/sources] 
            (* => nrel_format: format_github_source_link;; *)
        };; 
    *);;

// Full initiation condition of agent
..agent_calculate_set_power_initiation_condition
= [*
    action_calculate_set_power _-> .._action;;
    action_initiated _-> .._action;;
    .._action _-> rrel_1:: .._parameter;;
*];; 
// Agent should check by this template that initiated action is instance of 
// class `action_calculate_set_power` and that it has argument.

// Full result condition of agent
..agent_calculate_set_power_result_condition
= [*
    action_calculate_set_power _-> .._action;;
    action_initiated _-> .._action;;
    action_finished_successfully _-> .._action;;
    .._action _-> rrel_1:: .._parameter;;

>     .._action _=> nrel_result:: .._result;;

Author: Fixed

    Define action class as keynode in agent class or keynodes class.

!!! warning
    You must override methods `GetActionClass` and `DoProgram`.

You can specify any existing event types as a template argument to the `ScAgent` class. For example, you can create agent that will be triggered to sc-event of removing sc-element.

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

class MyAgent : public ScAgent<ScEventBeforeEraseElement>
{
public:
  ScAddr GetActionClass() const override;
  ScResult DoProgram(
    ScEventBeforeEraseElement const & event, ScAction & action) override;

  // Other user-defined methods.
};
```

If you want to change specification of this agent in knowledge base, then write like this:

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

class MyAgent : public ScBaseAgent
{
public:
  ScAddr GetActionClass() const override;
  ScResult DoProgram(
    ScElementaryEvent const & event, ScAction & action) override;

  // Other user-defined methods.
};
```

This implementation allows to provide any sc-event type to `DoProgram`.

!!! note
    `ScBaseAgent` is alias for `ScAgent<ScElementaryEvent>`.

!!! note

>     If you provide specification of your agent in knowledge base, then you don't need to override `GetActionClass`. See [**C++ Modules API**] to learn how to implement agents with specification in the knowledge base.

Reviewer: link does not work

### **ScActionInitiatedAgent**

In multi-agent systems most of the agents are implemented to execute actions initiated by other agents. While `ScAgent` is useful to create broad event handling logic, using it to handle action initiations requires some boilerplate. We've implemented another agent class to make it easier for our users to implement action-executing agents. These agents are named action agents. Implementing these agents requires passing action class node rather than checking initiation condition manually.

This class can be only used for agents that should be triggered by generating an output sc-arc from `action_initiated` class node.

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

// Inherit your agent class from `ScActionInitiatedAgent` class.
class MyAgent : public ScActionInitiatedAgent
{
public:
  // Here you should specify class of actions which the given agent performs. 
  // This overriding is required.
  ScAddr GetActionClass() const override;
  // Here you should implement program of the given agent. 
  // This overriding is required.

>   ScResult DoProgram(ScActionEvent const & event, ScAction & action) override;

Reviewer: shouldn't it now be called ScActionInitiatedEvent?

Author: fixed

This class can be used for all types of platform-dependent agents. The example using this class is represented below.

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

// Inherit your agent class from `ScAgent` class and specify template argument 
// as sc-event class. Here `ScEventAfterGenerateIncomingArc<
// ScType::EdgeAccessConstPosPerm>` is type of event to which the given 
// agent reacts.
class MyAgent : public ScAgent<
  ScEventAfterGenerateIncomingArc<ScType::EdgeAccessConstPosPerm>>
{
public:
  // Here you should specify class of actions which the given agent performs. 
  // Here `GetActionClass` overrides `GetActionClass` 
  // in `ScAgent` class. This overriding is required.
  ScAddr GetActionClass() const override;
  // Here you should implement program of the given agent. 
  // This overriding is required.
  ScResult DoProgram(
    ScEventAfterGenerateIncomingArc<
      ScType::EdgeAccessConstPosPerm> const & event, 

>     ScAction & action) override;

Reviewer: is there somewhere a mention of what is passed as action for different events?

Author: Described

!!! warning
    You must override methods `GetActionClass` and `DoProgram`.

You can specify any existing event types as a template argument to the `ScAgent` class. For example, you can create agent that will be triggered to sc-event of removing sc-element.

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

class MyAgent : public ScAgent<ScEventBeforeEraseElement>
{
public:
  ScAddr GetActionClass() const override;
  ScResult DoProgram(
    ScEventBeforeEraseElement const & event, ScAction & action) override;

  // Other user-defined methods.
};
```

If you want to change specification of this agent in knowledge base, then write like this:

```cpp
// File my_agent.hpp
#pragma once

#include <sc-memory/sc_agent.hpp>

class MyAgent : public ScElementaryEventAgent
{
public:
  ScAddr GetActionClass() const override;
  ScResult DoProgram(
    ScElementaryEvent const & event, ScAction & action) override;

  // Other user-defined methods.
};
```

This implementation allows to provide any sc-event type to `DoProgram`.

!!! note
    `ScElementaryEventAgent` is alias for `ScAgent<ScElementaryEvent>`.

!!! note

>     If you provide specification of your agent in knowledge base, then you don't need to override `GetActionClass`. See [**C++ Modules API**](modules.md) to learn how to implement agents with specification in the knowledge base.

Author: Fixed

### **ScActionInitiatedAgent**

> In multi-agent systems most of the agents are implemented to execute actions initiated by other agents. While `ScAgent` is useful to create broad event handling logic, using it to handle action initiations requires some boilerplate. We've implemented another agent class to make it easier for our users to implement action-executing agents. These agents are named action agents. Implementing these agents requires passing action class node rather than checking initiation condition manually.

Author: I propose to remove this sentence

> In multi-agent systems most of the agents are implemented to execute actions initiated by other agents. While `ScAgent` is useful to create broad event handling logic, using it to handle action initiations requires some boilerplate. We've implemented another agent class to make it easier for our users to implement action-executing agents. These agents are named action agents. Implementing these agents requires passing action class node rather than checking initiation condition manually.

Author: Or use ScActionInitiatedAgent instead

> In multi-agent systems most of the agents are implemented to execute actions initiated by other agents. While `ScAgent` is useful to create broad event handling logic, using it to handle action initiations requires some boilerplate. We've implemented another agent class to make it easier for our users to implement action-executing agents. These agents are named action agents. Implementing these agents requires passing action class node rather than checking initiation condition manually.

Author: Removed

  ScActionInitiatedEvent const & event, ScAction & action)
{
  // You can use object of ScAction as object of ScAddr.
  ScIterator3Ptr const it3 = m_context.Iterator3(action, ..., ...);

  // Some logic...

  return action.FinishSuccessfully();
}
```

##### Handling action result

```cpp
ScResult MyAgent::DoProgram(
  ScActionInitiatedEvent const & event, ScAction & action)
{
  // Some logic...
 
  action.FormResult(foundAddr1, generatedAddr1, ...); 
  // Or you can use `UpdateResult` method.
  return action.FinishSuccessfully();
}
```

To learn more about methods of `ScAction` class, see [**C++ Actions API**](actions.md).

```cpp
ScResult MyAgent::DoProgram(
  ScActionInitiatedEvent const & event, ScAction & action)
{
  // Some logic...
 
  action.SetResult(structureAddr);
  return action.FinishSuccessfully();
}
```

##### Handling action finish state

```cpp
ScResult MyAgent::DoProgram(
  ScActionInitiatedEvent const & event, ScAction & action)
{
  // Some logic...

  if (/* case 1 */)
    return action.FinishSuccessfully();
  else if (/* case 2 */)
    return action.FinishUnsuccessfully();
  else
    return action.FinishWithError();
}
```

```cpp
ScResult MyAgent::Program(

>   ScActionInitiatedEvent const & event, ScAction & action)

Author: Fixed

{
  action.IsInitiated(); // result: true
  action.IsFinished(); // result: false
  action.IsFinishedSuccessfully(); // result: false

  // Some logic...

  return action.FinishSuccessfully();
}
```

!!! warning
    The API of `ScAction` provides other methods. Don't use `GetResult` for initiated but not finished action and `Initiate` for initiated or finished action.

---

### **Not required methods**

#### **GetAbstractAgent**

It searches abstract agent for agent of specified class. If the agent implementation for this agent class is not included in any abstract sc-agent, then `GetAbstractAgent` will throw `utils::ExceptionItemNotFound`.

> You can redefine this method in your agent class.

Author: Fixed
